src/python/zquantum/core/serialization_test.py

```diff
@@ -85,3 +87,74 @@ def test_zapata_encoder_can_handle_optimization_result():
     deserialized_result = json.loads(json.dumps(result_to_serialize, cls=ZapataEncoder))
 
     assert deserialized_result == expected_deserialized_result
+
+
+def test_optimization_result_serialization_io():
+    # The result constructed below does not make sense.
+    # It does not matter though, as we are only testing serialization and it contains variety
+    # of data to be serialized.
+    result_to_serialize = optimization_result(
+        opt_value=0.5,
+        opt_params=np.array([0, 0.5, 2.5]),
+        nit=3,
+        fev=10,
+        history=[
+            HistoryEntry(
+                call_number=0,
+                params=np.array([0.1, 0.2, 0.3j]),
+                value=ValueEstimate(0.5, precision=6),
+            ),
+            HistoryEntry(call_number=1, params=np.array([1, 2, 3]), value=-10.0),
+            HistoryEntryWithArtifacts(
+                call_number=2,
+                params=np.array([-1, -0.5, -0.6]),
+                value=-20.0,
+                artifacts={
+                    "bitstring": "0111",
+                    "bitstring_distribution": BitstringDistribution(
+                        {"111": 0.25, "010": 0.75}
+                    ),
+                },
+            ),
+        ],
+    )
+
+    expected_deserialized_result = {
+        "schema": SCHEMA_VERSION + "-optimization_result",
+        "opt_value": 0.5,
+        "opt_params": convert_array_to_dict(np.array([0, 0.5, 2.5])),
+        "nit": 3,
+        "fev": 10,
+        "history": [
+            {
+                "call_number": 0,
+                "params": convert_array_to_dict(np.array([0.1, 0.2, 0.3j])),
+                "value": ValueEstimate(0.5, precision=6).to_dict(),
+            },
+            {
+                "call_number": 1,
+                "params": convert_array_to_dict(np.array([1, 2, 3])),
+                "value": -10.0,
+            },
+            {
+                "call_number": 2,
+                "params": convert_array_to_dict(np.array([-1, -0.5, -0.6])),
+                "value": -20.0,
+                "artifacts": {
+                    "bitstring": "0111",
+                    "bitstring_distribution": {"111": 0.25, "010": 0.75},
+                },
+            },
+        ],
+    }
+
+    optimization_result_filename = "test-optimization-result-io.json"
+
+    # When
+    save_optimization_results(result_to_serialize, optimization_result_filename)
+
+    # Then
+    with open(optimization_result_filename, "r") as f:
+        loaded_data = json.loads(f.read())
```

Comment:
Let's use `json.load(f)` instead of `json.loads(f.read())`.

As a side note, it is not the best idea to use the actual filesystem when unit testing, but I'll explain more about it when we have some time to chat and it would require some rework on our approach for writing serialization functions.

Reply:
Addressed here: 13e53ba